DeploymentException not thrown when using @ConfigProperties

`@ConfigProperties` does not behave the same as `@ConfigProperty`

When using `@ConfigProperty` to inject configuration, if the property is missing, a DeploymentException is thrown at start-up.

For example, adding the following to a Controller:
```
@ConfigProperty(name = "cassandra.connection.node") public String node;
@ConfigProperty(name = "cassandra.connection.port") public int port;
```
and the following to application.properties (notice additional r on node) :
```
cassandra.connection.noder=0.0.0.0
cassandra.connection.port=9042
```
results in an exception at start-up:
```
javax.enterprise.inject.spi.DeploymentException: No config value of type [java.lang.String] exists for: cassandra.connection.node
```

However, when using `@ConfigProperties` as follows:
```
@ConfigProperties(prefix = "cassandra")
public class CassandraConfiguration
{
  public ConnectionData connection;

  public static class ConnectionData
  {
    public String node;
    public int port;
  }
}
```
Given the same application.properties file as above, an exception is not thrown at start-up. Instead, the application starts, but throws an UnhandledException, which wraps a NoSuchElementException, when the property is accessed:
```
UnhandledException: java.lang.RuntimeException: Error injecting com.***.CassandraConfiguration com.***.cassandraConfiguration
....
Caused by: java.util.NoSuchElementException: Property cassandra.connection.node not found
```

`@ConfigProperties` works as expected when the application.properties are correct:
```
cassandra.connection.node=0.0.0.0
cassandra.connection.port=9042
```

--- a/docs/src/main/asciidoc/application-configuration-guide.adoc
+++ b/docs/src/main/asciidoc/application-configuration-guide.adoc
@@ -241,6 +241,9 @@ public interface GreetingConfiguration {
 <2> In this case since `name` was not set, the corresponding property will be `greeting.suffix`.
 <3> It is unnecessary to specify the `@ConfigProperty` annotation because the method name follows the getter method naming conventions (`greeting.name` being the corresponding property) and no default value is needed.
 
+When using `@ConfigProperties` on a class or an interface, if the value of one of its fields is not provided, the application startup will fail and a `javax.enterprise.inject.spi.DeploymentException` indicating the missing value information will be thrown.
+This does not apply to `Optional` fields and fields with a default value.
+
 === Additional notes on @ConfigProperties
 
 When using a regular class annotated with `@ConfigProperties` the class doesn't necessarily have to declare getters and setters.

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/configproperties/ClassConfigPropertiesUtil.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/configproperties/ClassConfigPropertiesUtil.java
@@ -1,15 +1,14 @@
 package io.quarkus.arc.deployment.configproperties;
 
-import static io.quarkus.arc.deployment.configproperties.ConfigPropertiesUtil.ReadOptionalResponse;
 import static io.quarkus.arc.deployment.configproperties.ConfigPropertiesUtil.createReadMandatoryValueAndConvertIfNeeded;
 import static io.quarkus.arc.deployment.configproperties.ConfigPropertiesUtil.createReadOptionalValueAndConvertIfNeeded;
 import static io.quarkus.arc.deployment.configproperties.ConfigPropertiesUtil.determineSingleGenericType;
 
 import java.lang.reflect.Modifier;
+import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
-import java.util.NoSuchElementException;
 import java.util.Optional;
 import java.util.Set;
 
@@ -26,6 +25,9 @@ import org.jboss.jandex.IndexView;
 import org.jboss.jandex.MethodInfo;
 import org.jboss.jandex.Type;
 
+import io.quarkus.arc.deployment.ConfigPropertyBuildItem;
+import io.quarkus.arc.deployment.configproperties.ConfigPropertiesUtil.ReadOptionalResponse;
+import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.bean.JavaBeanUtil;
 import io.quarkus.deployment.util.HashUtil;
 import io.quarkus.gizmo.BranchResult;
@@ -112,8 +114,9 @@ final class ClassConfigPropertiesUtil {
     /**
      * @return true if the configuration class needs validation
      */
-    static boolean addProducerMethodForClassConfigProperties(ClassInfo configPropertiesClassInfo,
-            ClassCreator producerClassCreator, String prefixStr, IndexView applicationIndex) {
+    static boolean addProducerMethodForClassConfigProperties(ClassLoader classLoader, ClassInfo configPropertiesClassInfo,
+            ClassCreator producerClassCreator, String prefixStr, IndexView applicationIndex,
+            BuildProducer<ConfigPropertyBuildItem> configProperties) {
 
         if (!DotNames.OBJECT.equals(configPropertiesClassInfo.superName())) {
             throw new IllegalArgumentException(
@@ -164,9 +167,8 @@ final class ClassConfigPropertiesUtil {
                 configObjectClassStr, produceMethodParameterTypes)) {
             methodCreator.addAnnotation(Produces.class);
 
-            ResultHandle configObject = populateConfigObject(configPropertiesClassInfo,
-                    prefixStr,
-                    methodCreator, applicationIndex);
+            ResultHandle configObject = populateConfigObject(classLoader, configPropertiesClassInfo, prefixStr, methodCreator,
+                    applicationIndex, configProperties);
 
             if (needsValidation) {
                 createValidationCodePath(methodCreator, configObject, prefixStr);
@@ -195,11 +197,13 @@ final class ClassConfigPropertiesUtil {
         }
     }
 
-    private static ResultHandle populateConfigObject(ClassInfo configClassInfo,
-            String prefixStr,
-            MethodCreator methodCreator, IndexView applicationIndex) {
+    private static ResultHandle populateConfigObject(ClassLoader classLoader, ClassInfo configClassInfo, String prefixStr,
+            MethodCreator methodCreator, IndexView applicationIndex, BuildProducer<ConfigPropertyBuildItem> configProperties) {
         String configObjectClassStr = configClassInfo.name().toString();
         ResultHandle configObject = methodCreator.newInstance(MethodDescriptor.ofConstructor(configObjectClassStr));
+
+        // Fields with a default value will be removed from this list at the end of the method.
+        List<ConfigPropertyBuildItemCandidate> configPropertyBuildItemCandidates = new ArrayList<>();
 
         // For each field of the class try to pull it out of MP Config and call the corresponding setter
         List<FieldInfo> fields = configClassInfo.fields();
@@ -240,9 +244,8 @@ final class ClassConfigPropertiesUtil {
                             "Nested configuration class " + fieldTypeClassInfo + " must be public ");
                 }
 
-                ResultHandle nestedConfigObject = populateConfigObject(fieldTypeClassInfo,
-                        prefixStr + "." + field.name(),
-                        methodCreator, applicationIndex);
+                ResultHandle nestedConfigObject = populateConfigObject(classLoader, fieldTypeClassInfo,
+                        prefixStr + "." + field.name(), methodCreator, applicationIndex, configProperties);
                 createWriteValue(methodCreator, configObject, field, setter, useFieldAccess, nestedConfigObject);
 
             } else {
@@ -286,16 +289,6 @@ final class ClassConfigPropertiesUtil {
                                     MethodDescriptor.ofMethod(configObjectClassStr, getterName, fieldTypeStr),
                                     configObject);
                         }
-
-                        /*
-                         * If the getter value is null (meaning that no default was set)
-                         * throw the exception that MP Config would.
-                         * If the value is not null, then we don't need to do anything since the class will already
-                         * contain the default value the user intended to use
-                         */
-                        isPresentFalse.ifNull(defaultValue).trueBranch()
-                                .throwException(NoSuchElementException.class,
-                                        "Property " + fullConfigName + " not found");
                     } else {
                         /*
                          * In this case we want a missing property to cause an exception that we don't handle
@@ -307,9 +300,19 @@ final class ClassConfigPropertiesUtil {
                         createWriteValue(methodCreator, configObject, field, setter, useFieldAccess, setterValue);
 
                     }
+                    configPropertyBuildItemCandidates
+                            .add(new ConfigPropertyBuildItemCandidate(field.name(), fullConfigName, fieldTypeStr));
                 }
             }
         }
+
+        ConfigPropertyBuildItemCandidateUtil.removePropertiesWithDefaultValue(classLoader, configObjectClassStr,
+                configPropertyBuildItemCandidates);
+        for (ConfigPropertyBuildItemCandidate candidate : configPropertyBuildItemCandidates) {
+            configProperties
+                    .produce(new ConfigPropertyBuildItem(candidate.getConfigPropertyName(), candidate.getConfigPropertyType()));
+        }
+
         return configObject;
     }
 

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/configproperties/ConfigPropertiesBuildStep.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/configproperties/ConfigPropertiesBuildStep.java
@@ -19,11 +19,13 @@ import org.jboss.jandex.DotName;
 import org.jboss.jandex.IndexView;
 
 import io.quarkus.arc.config.ConfigProperties;
+import io.quarkus.arc.deployment.ConfigPropertyBuildItem;
 import io.quarkus.arc.deployment.GeneratedBeanBuildItem;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.builditem.ApplicationIndexBuildItem;
 import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
+import io.quarkus.deployment.builditem.DeploymentClassLoaderBuildItem;
 import io.quarkus.deployment.builditem.GeneratedClassBuildItem;
 import io.quarkus.deployment.builditem.RunTimeConfigurationDefaultBuildItem;
 import io.quarkus.gizmo.ClassCreator;
@@ -36,7 +38,9 @@ public class ConfigPropertiesBuildStep {
             ApplicationIndexBuildItem applicationIndex,
             BuildProducer<GeneratedClassBuildItem> generatedClasses,
             BuildProducer<GeneratedBeanBuildItem> generatedBeans,
-            BuildProducer<RunTimeConfigurationDefaultBuildItem> defaultConfigValues) {
+            BuildProducer<RunTimeConfigurationDefaultBuildItem> defaultConfigValues,
+            BuildProducer<ConfigPropertyBuildItem> configProperties,
+            DeploymentClassLoaderBuildItem deploymentClassLoader) {
         IndexView index = combinedIndex.getIndex();
         Collection<AnnotationInstance> instances = index.getAnnotations(DotNames.CONFIG_PROPERTIES);
         if (instances.isEmpty()) {
@@ -76,12 +80,12 @@ public class ConfigPropertiesBuildStep {
                 /*
                  * In this case we need to generate an implementation of the interface that for each interface method
                  * simply pulls data from MP Config and returns it.
-                 * The generated producer bean simply needs to return a instance of the generated class
+                 * The generated producer bean simply needs to return an instance of the generated class
                  */
 
                 String generatedClassName = InterfaceConfigPropertiesUtil.generateImplementationForInterfaceConfigProperties(
                         classInfo, nonBeansClassOutput, index, prefixStr,
-                        defaultConfigValues);
+                        defaultConfigValues, configProperties);
                 InterfaceConfigPropertiesUtil.addProducerMethodForInterfaceConfigProperties(producerClassCreator,
                         classInfo.name(), generatedClassName);
 
@@ -90,10 +94,9 @@ public class ConfigPropertiesBuildStep {
                  * In this case the producer method contains all the logic to instantiate the config class
                  * and call setters for value obtained from MP Config
                  */
-                boolean needsValidation = ClassConfigPropertiesUtil.addProducerMethodForClassConfigProperties(classInfo,
-                        producerClassCreator,
-                        prefixStr,
-                        applicationIndex.getIndex());
+                boolean needsValidation = ClassConfigPropertiesUtil.addProducerMethodForClassConfigProperties(
+                        deploymentClassLoader.getClassLoader(), classInfo, producerClassCreator, prefixStr,
+                        applicationIndex.getIndex(), configProperties);
                 if (needsValidation) {
                     configClassesThatNeedValidation.add(classInfo.name());
                 }

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/configproperties/ConfigPropertiesUtil.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/configproperties/ConfigPropertiesUtil.java
@@ -1,6 +1,5 @@
 package io.quarkus.arc.deployment.configproperties;
 
-import java.util.Collection;
 import java.util.Optional;
 import java.util.function.IntFunction;
 

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/configproperties/ConfigPropertyBuildItemCandidate.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/configproperties/ConfigPropertyBuildItemCandidate.java
@@ -1,0 +1,28 @@
+package io.quarkus.arc.deployment.configproperties;
+
+public class ConfigPropertyBuildItemCandidate {
+
+    private final String fieldName;
+
+    private final String configPropertyName;
+
+    private final String configPropertyType;
+
+    public ConfigPropertyBuildItemCandidate(String fieldName, String configPropertyName, String configPropertyType) {
+        this.fieldName = fieldName;
+        this.configPropertyName = configPropertyName;
+        this.configPropertyType = configPropertyType;
+    }
+
+    public String getFieldName() {
+        return fieldName;
+    }
+
+    public String getConfigPropertyName() {
+        return configPropertyName;
+    }
+
+    public String getConfigPropertyType() {
+        return configPropertyType;
+    }
+}

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/configproperties/ConfigPropertyBuildItemCandidateUtil.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/configproperties/ConfigPropertyBuildItemCandidateUtil.java
@@ -1,0 +1,71 @@
+package io.quarkus.arc.deployment.configproperties;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.lang.reflect.Modifier;
+import java.util.List;
+
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+
+public class ConfigPropertyBuildItemCandidateUtil {
+
+    /**
+     * This method inspects the {@code configClass} bytecode to identify all fields that have a default value set in the class
+     * constructor. These fields are removed from the {@link ConfigPropertyBuildItemCandidate} list because we don't want to
+     * throw an exception if no config property value was provided for them. There is no bytecode modification performed during
+     * this process.
+     */
+    public static void removePropertiesWithDefaultValue(ClassLoader classLoader, String configClass,
+            List<ConfigPropertyBuildItemCandidate> candidates) {
+        try (InputStream is = classLoader.getResourceAsStream(configClass.replace('.', '/') + ".class")) {
+            ClassReader configClassReader = new ClassReader(is);
+            configClassReader.accept(new ConfigClassVisitor(candidates), 0);
+        } catch (IOException e) {
+            throw new UncheckedIOException(configClass + " class reading failed", e);
+        }
+    }
+
+    private static class ConfigClassVisitor extends ClassVisitor {
+
+        private List<ConfigPropertyBuildItemCandidate> candidates;
+
+        private ConfigClassVisitor(List<ConfigPropertyBuildItemCandidate> candidates) {
+            super(Opcodes.ASM7);
+            this.candidates = candidates;
+        }
+
+        @Override
+        public MethodVisitor visitMethod(int access, String name, String descriptor, String signature, String[] exceptions) {
+            MethodVisitor superMethodVisitor = super.visitMethod(access, name, descriptor, signature, exceptions);
+            if (access == Modifier.PUBLIC && name.equals("<init>") && descriptor.equals("()V")) {
+                return new ConfigClassConstructorVisitor(superMethodVisitor, candidates);
+            }
+            return superMethodVisitor;
+        }
+    }
+
+    private static class ConfigClassConstructorVisitor extends MethodVisitor {
+
+        private List<ConfigPropertyBuildItemCandidate> candidates;
+
+        private ConfigClassConstructorVisitor(MethodVisitor superMethodVisitor,
+                List<ConfigPropertyBuildItemCandidate> candidates) {
+            super(Opcodes.ASM7, superMethodVisitor);
+            this.candidates = candidates;
+        }
+
+        @Override
+        public void visitFieldInsn(int opcode, String owner, String name, String descriptor) {
+            // For each instruction in the config class constructor that is setting a field value, we remove that field from
+            // the candidates list if the field was part it.
+            if (opcode == Opcodes.PUTFIELD) {
+                candidates.removeIf(candidate -> candidate.getFieldName().equals(name));
+            }
+            super.visitFieldInsn(opcode, owner, name, descriptor);
+        }
+    }
+}

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/configproperties/InterfaceConfigPropertiesUtil.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/configproperties/InterfaceConfigPropertiesUtil.java
@@ -12,6 +12,7 @@ import java.util.Set;
 import javax.enterprise.inject.Produces;
 
 import org.eclipse.microprofile.config.Config;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
 import org.jboss.jandex.AnnotationInstance;
 import org.jboss.jandex.AnnotationValue;
 import org.jboss.jandex.ClassInfo;
@@ -20,6 +21,7 @@ import org.jboss.jandex.IndexView;
 import org.jboss.jandex.MethodInfo;
 import org.jboss.jandex.Type;
 
+import io.quarkus.arc.deployment.ConfigPropertyBuildItem;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.bean.JavaBeanUtil;
 import io.quarkus.deployment.builditem.RunTimeConfigurationDefaultBuildItem;
@@ -59,8 +61,8 @@ final class InterfaceConfigPropertiesUtil {
     }
 
     static String generateImplementationForInterfaceConfigProperties(ClassInfo originalInterface, ClassOutput classOutput,
-            IndexView index, String prefixStr,
-            BuildProducer<RunTimeConfigurationDefaultBuildItem> defaultConfigValues) {
+            IndexView index, String prefixStr, BuildProducer<RunTimeConfigurationDefaultBuildItem> defaultConfigValues,
+            BuildProducer<ConfigPropertyBuildItem> configProperties) {
         Set<DotName> allInterfaces = new HashSet<>();
         allInterfaces.add(originalInterface.name());
         collectInterfacesRec(originalInterface, index, allInterfaces);
@@ -151,6 +153,10 @@ final class InterfaceConfigPropertiesUtil {
                                     fullConfigName, returnType,
                                     method.declaringClass().name(), methodCreator, config);
                             methodCreator.returnValue(value);
+                            if (defaultValueStr == null || ConfigProperty.UNCONFIGURED_VALUE.equals(defaultValueStr)) {
+                                configProperties
+                                        .produce(new ConfigPropertyBuildItem(fullConfigName, returnType.name().toString()));
+                            }
                         }
                     }
                 }

--- a/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/configproperties/ClassWithMissingConfigPropertiesTest.java
+++ b/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/configproperties/ClassWithMissingConfigPropertiesTest.java
@@ -1,0 +1,59 @@
+package io.quarkus.arc.test.configproperties;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.spi.DeploymentException;
+import javax.inject.Inject;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.arc.config.ConfigProperties;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class ClassWithMissingConfigPropertiesTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest TEST = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class).addClasses(Configured.class, DummyProperties.class))
+            .assertException(e -> {
+                assertEquals(DeploymentException.class, e.getClass());
+                assertTrue(e.getMessage().contains("dummy.foo"));
+            });
+
+    @Test
+    public void shouldNotBeInvoked() {
+        // This method should not be invoked
+        fail();
+    }
+
+    @ApplicationScoped
+    public static class Configured {
+
+        @Inject
+        DummyProperties dummyProperties;
+
+        public String getFoo() {
+            return dummyProperties.getFoo();
+        }
+    }
+
+    @ConfigProperties
+    public static class DummyProperties {
+
+        private String foo;
+
+        public String getFoo() {
+            return foo;
+        }
+
+        public void setFoo(String foo) {
+            this.foo = foo;
+        }
+    }
+}

--- a/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/configproperties/ClassWithMissingNestedConfigPropertiesTest.java
+++ b/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/configproperties/ClassWithMissingNestedConfigPropertiesTest.java
@@ -1,0 +1,56 @@
+package io.quarkus.arc.test.configproperties;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.spi.DeploymentException;
+import javax.inject.Inject;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.arc.config.ConfigProperties;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class ClassWithMissingNestedConfigPropertiesTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest TEST = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class).addClasses(Configured.class, DummyProperties.class))
+            .assertException(e -> {
+                assertEquals(DeploymentException.class, e.getClass());
+                assertTrue(e.getMessage().contains("dummy.inner.foo"));
+            });
+
+    @Test
+    public void shouldNotBeInvoked() {
+        // This method should not be invoked
+        fail();
+    }
+
+    @ApplicationScoped
+    public static class Configured {
+
+        @Inject
+        DummyProperties dummyProperties;
+
+        public String getFoo() {
+            return dummyProperties.inner.foo;
+        }
+    }
+
+    @ConfigProperties
+    public static class DummyProperties {
+
+        public InnerProperties inner;
+
+        public static class InnerProperties {
+
+            public String foo;
+        }
+    }
+}

--- a/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/configproperties/InterfaceNoGetterWithMissingConfigPropertiesTest.java
+++ b/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/configproperties/InterfaceNoGetterWithMissingConfigPropertiesTest.java
@@ -1,0 +1,53 @@
+package io.quarkus.arc.test.configproperties;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.spi.DeploymentException;
+import javax.inject.Inject;
+
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.arc.config.ConfigProperties;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class InterfaceNoGetterWithMissingConfigPropertiesTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest TEST = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class).addClasses(Configured.class, DummyProperties.class))
+            .assertException(e -> {
+                assertEquals(DeploymentException.class, e.getClass());
+                assertTrue(e.getMessage().contains("dummy.foo"));
+            });
+
+    @Test
+    public void shouldNotBeInvoked() {
+        // This method should not be invoked
+        fail();
+    }
+
+    @ApplicationScoped
+    public static class Configured {
+
+        @Inject
+        DummyProperties dummyProperties;
+
+        public String getFoo() {
+            return dummyProperties.foo();
+        }
+    }
+
+    @ConfigProperties
+    public interface DummyProperties {
+
+        @ConfigProperty(name = "foo")
+        String foo();
+    }
+}

--- a/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/configproperties/InterfaceWithMissingConfigPropertiesTest.java
+++ b/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/configproperties/InterfaceWithMissingConfigPropertiesTest.java
@@ -1,0 +1,51 @@
+package io.quarkus.arc.test.configproperties;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.spi.DeploymentException;
+import javax.inject.Inject;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.arc.config.ConfigProperties;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class InterfaceWithMissingConfigPropertiesTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest TEST = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class).addClasses(Configured.class, DummyProperties.class))
+            .assertException(e -> {
+                assertEquals(DeploymentException.class, e.getClass());
+                assertTrue(e.getMessage().contains("dummy.foo"));
+            });
+
+    @Test
+    public void shouldNotBeInvoked() {
+        // This method should not be invoked
+        fail();
+    }
+
+    @ApplicationScoped
+    public static class Configured {
+
+        @Inject
+        DummyProperties dummyProperties;
+
+        public String getFoo() {
+            return dummyProperties.getFoo();
+        }
+    }
+
+    @ConfigProperties
+    public interface DummyProperties {
+
+        String getFoo();
+    }
+}

